### ISSUE TITLE
Reverse lookup draft 

### ISSUE THREAD

**Issue by tcezard** [ga4gh/reverse-lookup-spec#1]
As discussed during the call I'm opening up the initial issue for discussing what a draft specification would look like 

I'll give it a first stab. 
## Use case 

The main use case is to retrieve the digest(s) corresponding to a sequence from a set of metadata to enable user to find out the refget digest of a sequence without having to calculate it or even access the sequence.
A google doc exists with User stories (link is in the call's minutes)  so please add to it if you have any other use cases.

## Endpoint name
It was raised as something that needed to be clarified at the last call
I would propose `refget_lookup` to make clear that it is related to the [refget](https://github.com/samtools/hts-specs/blob/master/refget.md) protocol. 
There was also the idea floating about last year that we could have two endpoints. One for unique queries and one for searches. 

## Filtering 

I would suggest that filtering is done via keyword argument with a controlled vocabulary. `name=chr1&assembly=GCA_000001405.15`
Another solution offered by @reece was to use CURIE to define the namespace GCA_000001405.15:chr1 but this limits the filtering to identifiers rather than any metadata.

If we decide to go down the route of using two end point we could imagine both options being used: 
 - The unique endpoint would support a single query argument built as a CURIE with the first element providing the namespace and second the identifier. I still see two competing options for the namespace: 
   - The assembly build `GCA_000001405.15` or any other namespace that makes the second part of the CURIE unique
   - The naming authority: `insdc` `ensembl` or `refseq` that do make the query more precise but reduce the search space
 - For the search endpoint, the filtering would be done via keyword argument with a controlled vocabulary. `name=chr1&assembly=GCA_000001405.15`

## Controlled vocabulary 
The fields we have talked about so far are 
 - name: Given name of the sequence 
 - length: length of the sequence
 - assembly: assembly this sequence was included in
 - taxonomy id: taxonomy id of the species the sequence was found in
 - build: the group of sequences where this sequence was defined
 - alphabet: type of alphabet used in the sequence

## Returned Payload 
List of records that match the query filtered paginated if necessary
``` json
[
  {
    "md5": "md5 of the sequence",
    "trunc512": "trunc512 of the sequence"
  },
 {

 }
] 
```

@andrewyatz, @raskoleinonen, @daviesrob Feel free to add, remove, or comment

**Comment by tcezard** [comment]
One thing we could do is to start with a simple endpoint that only accept one argument like `identifier` the identifier could be an accession (INSDC, Refseq, Ensembl, etc) or a CURIE  `GCA_000001405.15:chr1` or whatever the implementation would support. 

We could then extend to other attribute at a later date.

**Comment by andrewyatz** [comment]
If an identifier is unique, then much like Reece's CURIE scheme refget could already support this. But that's more about name-spacing of the identifier using identifiers.org i.e. the `GCA_000001405.15:chr1` wouldn't be a valid identifier. However that doesn't mean it's not a valid CURIE as so long as you can inject the right hand side into an appropriate URL (as defined by the left hand side) then it should resolve (see [CURIE spec](https://www.w3.org/TR/2010/NOTE-curie-20101216/)).

I would say the controlled vocab is very much tuned to the use case of assemblies, but that's not a bad thing. More we'd need to accept that fact. Also considering attributes is the hardest thing to agree on (see ga4gh/seqcol-spec#8) then I'd be okay with the scope being limited.

Last point is I would push for GA4GH identifier over trunc512. We need to retire trunc512 ideally ... even though it is compatible with GA4GH identifier

**Comment by andrewyatz** [comment]
> One thing we could do is to start with a simple endpoint that only accept one argument like `identifier` the identifier could be an accession (INSDC, Refseq, Ensembl, etc) or a CURIE `GCA_000001405.15:chr1` or whatever the implementation would support.
> 
> We could then extend to other attribute at a later date.

Speaking in today's meeting a simple identifier would be the right way forward here. It is our simplest use-case and easier to implement. Reverse lookup should allow for an ambiguous input (even an identifier) and give multiple results back. A more exact input should give one "hit" but the payload should always support multiple hits.

Refget "classic" can be adapted to behave just the same way Reece's sequence retrieval service works in accepting an identifier in and giving just a single response. Refget is already built around the idea the questions have to be exact. This would still be in refget's spirit. The spec currently says the `<id>` parameter is `A string specifying an identifier to retrieve sequence for using one of the defined checksum algorithms or a server-specific checksum algorithm`. That can be weakened to something more like `server-specific identifier system`. 

In summary

- We should go for the simplest case
- A multiple hit payload response is correct
- A single hit payload is the purpose of refget

**Comment by raskoleinonen** [comment]
I second @andrewyatz suggestion to use refget ids, however, I would further suggest that the ids should only return a single response (like INSDC sequence versions would) i.e. I would define the <id> parameter something like:

```A string specifying an identifier uniquely associated with a sequence so that at most one sequence is returned for an id. The id can be a checksum or other unique identifier supported by the server```. 

I would expect the server to return 500 Internal Server Error if for some reason more than one sequence would be returned for a given id.

ENA's CRAM reference registry could provide a reference implementation for this using INSDC sequence versions fairly quickly.

I would implement more complex searches and support for multiple hits as part of different specs (```refget_lookup```).

**Comment by andrewyatz** [comment]
Good point @raskoleinonen and yes I think you're right more than one sequence coming back is an issue for the server so a 500 would be appropriate. Going through this flow chart https://www.codetinkerer.com/2015/12/04/choosing-an-http-status-code.html I think you're right.
